### DISCUSSION TITLE
[QUESTION] Is aiokafka consumer thread safe?

### DISCUSSION THREAD

**Discussion by DavidOsparks** [aio-libs/aiokafka#959] [Q&A]
**confluent-kafka-python** says _"Producer, Consumer and AdminClient are all thread safe."_
**kafka-python** says _"while it is possible to use the KafkaConsumer in a thread-local manner, multiprocessing is recommended."_
However, I cannot find any description about thread safety in the document of [aiokafka](https://aiokafka.readthedocs.io/en/stable/), so I worry about if AIOKafkaConsumer is thread safe? Whether it is or not, maybe it should be mentioned in the document.

**Comment by vmaurin** [comment]
In python, all the "aio/async" libraries run over an event loop. An event loop is always run by a single thread, managing multiple concurrent queries. So the first sentence of the documentation
> aiokafka is a client for the Apache Kafka distributed stream processing system using [asyncio](http://docs.python.org/3.7/library/asyncio.html).

should be enough to indicate about the threading model

**Reply by antonagestam** [reply]
Not really. A piece of code designed for asyncio does not by design either support or not being run across multiple threads.

**Reply by DavidOsparks** [reply]
Thanks. I checked some docs, maybe we can say that:

1. asyncio is not totally thread-safe. In fact, it is almost totally not thread-safe.
2. But under some conditions, we can still use asyncio with multithreading.

[https://docs.python.org/3/library/asyncio-dev.html#concurrency-and-multithreading](https://docs.python.org/3/library/asyncio-dev.html#concurrency-and-multithreading)
"Almost all asyncio objects are not thread safe, which is typically not a problem unless there is code that works with them from outside of a Task or a callback."

[https://docs.python.org/3/library/asyncio-sync.html](https://docs.python.org/3/library/asyncio-sync.html)
"asyncio primitives are not thread-safe, therefore they should not be used for OS thread synchronization (use [threading](https://docs.python.org/3/library/threading.html#module-threading) for that)"

However, back to original question: is aiokafka consumer thread-safe? I have no idea if aiokafka uses asyncio (or some other things) unsafely.
